common/stacks/src/main/scala/weco/api/stacks/services/ItemLookup.scala

```diff
@@ -0,0 +1,106 @@
+package weco.api.stacks.services
+
+import com.sksamuel.elastic4s.ElasticDsl._
+import com.sksamuel.elastic4s.{ElasticError, Index}
+import uk.ac.wellcome.models.Implicits._
+import weco.api.search.elasticsearch.ElasticsearchService
+import weco.catalogue.internal_model.identifiers.{
+  CanonicalId,
+  IdState,
+  SourceIdentifier
+}
+import weco.catalogue.internal_model.work.{Item, Work}
+import weco.catalogue.internal_model.work.WorkState.Indexed
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class ItemLookup(elasticsearchService: ElasticsearchService)(
+  implicit ec: ExecutionContext
+) {
+
+  /** Returns an Item that corresponds to this canonical ID.
+    *
+    * Note: the output from this function may not be consistent, even against
+    * a fixed API index.  Because of merging logic, some item IDs have varying
+    * representations on different works.
+    *
+    */
+  def byCanonicalId(itemId: CanonicalId)(index: Index)
+    : Future[Either[ElasticError, Option[Item[IdState.Identified]]]] = {
+    val searchRequest =
+      search(index)
+        .query(
+          boolQuery.filter(
+            termQuery("data.items.id.canonicalId", itemId.underlying),
+            termQuery(field = "type", value = "Visible")
+          )
+        )
+        .size(1)
+
+    elasticsearchService.findBySearch[Work[Indexed]](searchRequest).map {
+      case Left(err) => Left(err)
+      case Right(works) =>
+        Right(
+          works
+            .flatMap { _.data.items }
+            .collectFirst {
+              // The .asInstanceOf here will be a no-op at runtime, and is just so
+              // the compiler knows this will always be an Item[IdState.Identified]
+              case item @ Item(IdState.Identified(id, _, _), _, _)
+                  if id == itemId =>
+                item.asInstanceOf[Item[IdState.Identified]]
+            }
+        )
+    }
+  }
+
+  /** Returns an Item that corresponds to this source identifier.
+    *
+    * Note: the output from this function may not be consistent, even against
+    * a fixed API index.  Because of merging logic, some item IDs have varying
+    * representations on different works.
+    *
+    */
+  def bySourceIdentifier(itemId: SourceIdentifier)(index: Index)
```

Comment:
How useful is this if it cannot be consistent? If it's intended use is in the items & requesting services will an inconsistent lookup suffice?

Reply:
So it depends on how much information we want to expose here – we could make it consistent by tightening the output.

You want to go canonical ID ~> Sierra source identifier, so you can talk to Sierra.

You want to go Sierra source identifier ~> item canonical ID, possibly work ID (although this is a little tricky wrt Works because it's not 1:1).

If we clamped these methods to just return IDs (making them more of an IdentifierLookup), that would I think be consistent and might be better?

Reply:
> If we clamped these methods to just return IDs (making them more of an IdentifierLookup), that would I think be consistent and might be better?

Would that disrupt who you might have intended to use the work lookup?

Reply:
I don't *think* so – the Work lookup would be fine as-is, because it does return unambiguous results.

It's just the Items lookup that would be affected. It would mean we can't expose any other item information in the list of results; whether that's an issue depends on how much we want to return. (We don't currently put much info in the item results: https://github.com/wellcomecollection/catalogue-api/blob/main/requests/src/test/scala/uk/ac/wellcome/platform/api/requests/RequestsApiFeatureTest.scala#L121-L147